internal/querywatcher/query_watcher.go

```diff
@@ -163,35 +187,56 @@ func (w *QueryWatcher) updateQueryCache(query *DSQLQuery, event dstore.WatchEven
 	}
 }
 
-// notifyClients notifies all clients watching a query about the new result.
-func (w *QueryWatcher) notifyClients(query *DSQLQuery, clients *sync.Map, queryResult *[]dstore.DSQLQueryResultRow) {
+// notifyClient notifies all clients watching a query about the new result.
+func (w *QWatchManager) notifyClient(query *DSQLQuery, clientFD int, queryResult *[]dstore.DSQLQueryResultRow) {
 	encodedResult := clientio.Encode(CreatePushResponse(query, queryResult), false)
-	clients.Range(func(clientKey, _ interface{}) bool {
-		clientFD := clientKey.(int)
-		_, err := syscall.Write(clientFD, encodedResult)
-		if err != nil {
-			w.removeWatcher(query, clientFD)
-		}
-		return true
-	})
+	// Send the result to the client
+	_, err := syscall.Write(clientFD, encodedResult)
+	if err != nil {
+		w.removeWatcher(query, clientFD)
+	}
 }
 
 // serveAdhocQueries listens for adhoc queries, executes them, and sends the result back to the client.
-func (w *QueryWatcher) serveAdhocQueries(ctx context.Context) {
+func (w *QWatchManager) serveAdhocQueries(ctx context.Context) {
 	for {
 		select {
 		case query := <-AdhocQueryChan:
 			result, err := w.runQuery(&query.Query)
-			query.ResponseChan <- AdhocQueryResult{Result: result, Err: err}
+			query.ResponseChan <- AdhocQueryResult{
+				Result:      result,
+				Fingerprint: generateQueryFingerprint(&query.Query),
+				Err:         err,
+			}
 		case <-ctx.Done():
 			return
 		}
 	}
 }
 
+// generateQueryFingerprint used to creating a common fingerprint for similar queries
+func generateQueryFingerprint(query *DSQLQuery) string {
+	hash := sha256.Sum256([]byte(query.KeyRegex))
+	fingerprint := hex.EncodeToString(hash[:])
+	return "f_" + fingerprint
+}
+
 // addWatcher adds a client as a watcher to a query.
-func (w *QueryWatcher) addWatcher(query *DSQLQuery, clientFD int, cacheChan chan *[]dstore.KeyValue) {
-	clients, _ := w.WatchList.LoadOrStore(*query, &sync.Map{})
+func (w *QWatchManager) addWatcher(query *DSQLQuery, clientFD int, cacheChan chan *[]dstore.KeyValue) {
+	queryFingerprint := generateQueryFingerprint(query)
```

Comment:
+1 for this.
Also please correct me if I'm missing something, for similar queries currently we're just taking `tableName` into account. We can also take into account where clauses/selections to group similar queries together correct? Would that be an iteration on this PR?

Reply:
The table name is the core factor in deciding which keys lie in the result set. No need to further segregate them based on other clauses (since it takes more memory and offers minimal benefit)

Reply:
We can refine the fingerprinting logic in later iterations; the table name provides an excellent common factor across multiple queries. However, it can also be a subset of the table name later, if we want to have that.